File: backend/app/document_types.py
"""Loads the shared document-type catalog (repo-root catalog.json).

catalog.json's `documentTypes` array is the single source of truth for which
fields and party roles each of the 11 supported document types has. The
frontend reads the same file at build time (see
frontend/lib/document-types.ts) -- both sides parse one shared file instead
of hand-authoring a schema per document type on each side.
"""

import json
from functools import lru_cache
from pathlib import Path
from typing import Literal

from fastapi import HTTPException
from pydantic import BaseModel

CATALOG_FILE = Path(__file__).resolve().parents[2] / "catalog.json"

FieldKind = Literal["text", "textarea", "choice"]
Renderer = Literal["nda-coverpage", "generic-keyterms"]


class FieldChoice(BaseModel):
    value: str
    label: str


class FieldSpec(BaseModel):
    key: str
    label: str
    kind: FieldKind = "text"
    helpText: str = ""
    default: str = ""
    choices: list[FieldChoice] = []


class PartyRoleSpec(BaseModel):
    key: str
    label: str


class DocumentTypeSpec(BaseModel):
    slug: str
    name: str
    description: str
    renderer: Renderer
    standardTermsFilename: str
    coverPageFilename: str | None = None
    parties: list[PartyRoleSpec]
    fields: list[FieldSpec]


@lru_cache
def _document_types_by_slug() -> dict[str, DocumentTypeSpec]:
    catalog = json.loads(CATALOG_FILE.read_text())
    specs = [DocumentTypeSpec.model_validate(entry) for entry in catalog["documentTypes"]]
    return {spec.slug: spec for spec in specs}


def list_document_types() -> list[DocumentTypeSpec]:
    return list(_document_types_by_slug().values())


def get_document_type(slug: str) -> DocumentTypeSpec:
    try:
        return _document_types_by_slug()[slug]
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Unknown document type: {slug}") from None

File: frontend/lib/document-types.ts
import fs from "node:fs/promises";
import path from "node:path";

/**
 * Mirrors backend/app/document_types.py's DocumentTypeSpec. Both sides parse
 * the same repo-root catalog.json (see PL-9) rather than hand-authoring a
 * schema per document type on each side.
 */

export type FieldKind = "text" | "textarea" | "choice";

export interface FieldChoice {
  value: string;
  label: string;
}

export interface FieldSpec {
  key: string;
  label: string;
  kind: FieldKind;
  helpText?: string;
  default?: string;
  choices?: FieldChoice[];
}

export interface PartyRoleSpec {
  key: string;
  label: string;
}

export type DocumentRenderer = "nda-coverpage" | "generic-keyterms";

export interface DocumentTypeSpec {
  slug: string;
  name: string;
  description: string;
  renderer: DocumentRenderer;
  standardTermsFilename: string;
  coverPageFilename?: string;
  parties: PartyRoleSpec[];
  fields: FieldSpec[];
}

interface Catalog {
  documentTypes: DocumentTypeSpec[];
}

// Repo root is one level up from the Next.js app (see PL-5), matching the
// path convention app/page.tsx already uses for reading templates/.
function repoRootPath(...segments: string[]): string {
  return path.join(process.cwd(), "..", ...segments);
}

async function readCatalog(): Promise<Catalog> {
  const raw = await fs.readFile(repoRootPath("catalog.json"), "utf8");
  return JSON.parse(raw) as Catalog;
}

export async function listDocumentTypes(): Promise<DocumentTypeSpec[]> {
  const catalog = await readCatalog();
  return catalog.documentTypes;
}

export async function getDocumentType(slug: string): Promise<DocumentTypeSpec> {
  const spec = (await listDocumentTypes()).find((type) => type.slug === slug);
  if (!spec) throw new Error(`Unknown document type: ${slug}`);
  return spec;
}

export async function readTemplateFile(relativeFilename: string): Promise<string> {
  return fs.readFile(repoRootPath(relativeFilename), "utf8");
}
